src/jit/lowerxarch.cpp

```diff
@@ -1174,13 +1174,13 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
     else
     {
         MakeSrcContained(tree, shiftBy);
-    }
 
-    // Codegen of shift oper sets ZF and SF flags.
-    // Note that Rotate Left/Right instructions don't set ZF and SF flags.
-    if (tree->OperIsShift())
-    {
-        tree->gtFlags |= GTF_ZSF_SET;
+        // Codegen of shift oper sets ZF and SF flags provided shift count is not zero.
+        // Note that Rotate Left/Right instructions don't set ZF and SF flags.
+        if (tree->OperIsShift() && (shiftBy->gtIntConCommon.IconValue() != 0))
+        {
+            tree->gtFlags |= GTF_ZSF_SET;
+        }
```

Comment:
Yes, it's rather surprising. Also, even though I'm guessing that the test case reflects the original bug case, it would be good to test the constant cases as well.

---
In reply to: [90961146](https://github.com/dotnet/coreclr/pull/8461#discussion_r90961146) [](ancestors = 90961146)